extensions/sql-bindings/src/dialogs/addSqlBindingQuickpick.ts

```diff
@@ -45,60 +48,80 @@ export async function launchAddSqlBindingQuickpick(uri: vscode.Uri | undefined):
+			projectUri = await azureFunctionsUtils.getAFProjectContainingFile(uri);
+		} catch (e) {
+			// continue even if there's no AF project found. The binding should still be able to be added as long as there was an azure function found in the file earlier
+		}
 
-	// 4. ask for connection string setting name
-	let projectUri: vscode.Uri | undefined;
-	try {
-		projectUri = await azureFunctionsUtils.getAFProjectContainingFile(uri);
-	} catch (e) {
-		// continue even if there's no AF project found. The binding should still be able to be added as long as there was an azure function found in the file earlier
-	}
+		quickPickStep = 'updateConnectionString';
+		let connectionStringSettingName = await azureFunctionsUtils.promptAndUpdateConnectionStringSetting(projectUri);
+		if (!connectionStringSettingName) {
+			return;
+		}
+		TelemetryReporter.createActionEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.updateConnectionString)
+			.withAdditionalProperties(propertyBag).send();
 
-	let connectionStringSettingName = await azureFunctionsUtils.promptAndUpdateConnectionStringSetting(projectUri);
-	if (!connectionStringSettingName) {
-		return;
-	}
+		// 5. insert binding
+		try {
+			const result = await addSqlBinding(selectedBinding.type, uri.fsPath, azureFunctionName, objectName, connectionStringSettingName);
 
-	// 5. insert binding
-	try {
-		const result = await addSqlBinding(selectedBinding.type, uri.fsPath, azureFunctionName, objectName, connectionStringSettingName);
+			if (!result.success) {
+				void vscode.window.showErrorMessage(result.errorMessage);
+				TelemetryReporter.createErrorEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding)
+					.withAdditionalProperties(propertyBag).send();
+				return;
+			}
+			TelemetryReporter.createActionEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding)
+				.withAdditionalProperties(propertyBag).send();
 
-		if (!result.success) {
-			void vscode.window.showErrorMessage(result.errorMessage);
-			TelemetryReporter.sendErrorEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding);
+		} catch (e) {
+			void vscode.window.showErrorMessage(utils.getErrorMessage(e));
+			TelemetryReporter.createErrorEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding)
+				.withAdditionalProperties(propertyBag).send();
 			return;
 		}
-
-		TelemetryReporter.createActionEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding)
-			.withAdditionalProperties({ bindingType: selectedBinding.label })
-			.send();
-	} catch (e) {
-		void vscode.window.showErrorMessage(utils.getErrorMessage(e));
-		TelemetryReporter.sendErrorEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.finishAddSqlBinding);
-		return;
+	} finally {
+		propertyBag.quickPickStep = quickPickStep;
+		TelemetryReporter.createActionEvent(TelemetryViews.SqlBindingsQuickPick, TelemetryActions.exitSqlBindingsQuickpick)
```

Comment:
Right now if an exception occurs that isn't caught then we won't be able to tell when a user specifically exited the flow (cancel) or an error occurred. 

So would suggest looking into : 

1. Adding a catch here as well that sends an error event and then re-throws the error
2. Having a property `exitReason` that is initialized as `cancelled`, and then is updated in the catch to `error`, or to `done` at the end of the try then the user successfully completes